queries/highlights.scm

```diff
@@ -162,15 +168,244 @@
+          right: (identifier) @variable))))
+  ]))
+
+; type or class definitions and annotations
+
+((tuple_type "tuple" @type.builtin (field_declaration_list))
+ (#set! "priority" 101)) ; > @type
+; for inline tuple types
+
+; TODO: is there another way?
+(
+ [
+   (type_expression (identifier) @type.builtin) 
+   (type_expression (_ (identifier) @type.builtin)) 
+   (type_expression (_ (_ (identifier) @type.builtin))) 
+   (type_expression (_ (_ (_ (identifier) @type.builtin)))) 
+   (type_expression (_ (_ (_ (_ (identifier) @type.builtin))))) 
+   (type_expression (_ (_ (_ (_ (_ (identifier) @type.builtin)))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (identifier) @type.builtin))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin)))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin)))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin))))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin)))))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin))))))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin)))))))))))))) 
+ ]
+ (#set! "priority" 101) ; > @type
+ (#any-of? @type.builtin
+  "byte"
+  "cint"
+  "int"
+  "int8"
+  "int16"
+  "int32"
+  "int64"
+  "uint"
+  "uint8"
+  "uint16"
+  "uint32"
+  "uint64"
+  "float"
+  "float32"
+  "float64"
+  "bool"
+  "char"
+  "string"
+  "cstring"
+  "ref"
+  "ptr"
+  "range"
+  "array"
+  "seq"
+  "openArray"
+  "varargs"
+  "set"
+  "itarable"
+  "typedesc"
+  "untyped"
+  "auto"
+  "pointer"
+  "void"
+  "Rune"
+  "UncheckedArray"
+  "RootObj"
+  "SomeFloat"
+  "SomeInteger"
+  "SomeOrdinal"
+  "SomeNumber"
+  "SomeSignedInt"
+  "SomeUnsignedInt"
+  "Natural"
+  "Ordinal"
+  "Positive"
+ ))
+; built-in types
+
+(type_section ("type") @type.definition)
+; type definitions (e.g. `typedef` in C)
+
+(var_type "var" @type.qualifier)
+(out_type "out" @type.qualifier)
+(distinct_type "distinct" @type.qualifier)
+(ref_type "ref" @type.qualifier)
+(pointer_type "ptr" @type.qualifier)
+; type qualifiers (e.g. `const`)
+
+(tuple_construction
+  (colon_expression left: (_) @field))
+
+(field_declaration
+  (symbol_declaration_list
+    (symbol_declaration
+      name: (_) @field)))
+
+(dot_expression
+  right: (identifier) @field)
+; object and struct fields
+
+; TODO: is there another way?
+(proc_type (pragma_list [
+  (identifier) @variable
+  (_ (identifier) @variable)
+  (_ (_ (identifier) @variable))
+  (_ (_ (_ (identifier) @variable)))
+  (_ (_ (_ (_ (identifier) @variable))))
+  (_ (_ (_ (_ (_ (identifier) @variable)))))
+  (_ (_ (_ (_ (_ (_ (identifier) @variable))))))
+  (_ (_ (_ (_ (_ (_ (_ (identifier) @variable)))))))
+  (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @variable))))))))
+]) (#set! "priority" 101))
```

Comment:
Currently `type_expression`s are captured as a whole as `@type`.
That means everything within gets highlighted the same (yellow in my case).
Things that aren't supposed to get highlighted like a type, eg. punctuation,
need to be overruled.

Consider: `type a = proc() {.tags: [IO].}`

Without the above queries every identifier inside pragmas of proc types gets highlighted same as types, since nothing overrules them.
Would be great to have a way to express an arbitrary amount of nesting in the AST,
but afaik you can't with the query language, so this was the best I could do.
This catches identifiers up to 8 nodes deep.

Reply:
Then maybe we should opt for selective highlighting within `type_expression`.

The rules right now would make `module.T` and `seq[int]` looks ridiculous.

Reply:
> seq[int]
looks good for me. both `seq` and `int` get captured as @type.builtin, as intended

> module.T
indeed looks a little weird. are dot_expressions in types always module.type? There is also enum.field right? and probably other stuff like having typedesc in tuples and accessing them etc.

> Then maybe we should opt for selective highlighting within type_expression.

enlighten me :D

Reply:
> looks good for me. both `seq` and `int` get captured as @type.builtin, as intended

Since type_expression is yellow, wouldn't it made the brackets yellow, too? And a rule that broad at high priority (since it's positioned towards the end), would overrule `number_literal`, I imagine.

Here are some rules that I was playing with (and totally forgot to push):

```scheme
; Types
(_
  return_type: [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ])
(_
  type: [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ])
(_
  inherits: [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ])
(_
  refines: [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ])
(var_type
  [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ])
(ref_type
  [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ])
(distinct_type
  [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ])
(pointer_type
  [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ])
; Constructor
(call
  function: [
    (identifier) @type
    (accent_quoted (identifier) @type)
    (bracket_expression
      left: (identifier) @type
      right: (argument_list
        (identifier) @type)?)
  ]
  (argument_list
    (colon_expression)))

(type_symbol_declaration
  name: [
    (identifier) @type
    (exported_symbol
      (identifier) @type)
  ])
```

Though these don't penetrate well to lower levels, ie. `seq[Box[T]]`.

Reply:
> Since type_expression is yellow, wouldn't it made the brackets yellow, too?

Brackets get overruled by the `@punctuation.delimiter`

> And a rule that broad at high priority (since it's positioned towards the end), would overrule number_literal, I imagine.

No. I was surprised to find out, that there seems to be no more priority based on ordering in the file. I do not know what is responsible for choosing, which rules override which, if they have no priority specified. If they have an internal ordering for the capture groups or if the colorscheme now decides by applying certain highlights later than others.

![proof](https://github.com/alaviss/tree-sitter-nim/assets/36712072/f1531c7c-141e-48e1-a6f9-e0ea0a6ed679)
(nvim v0.9.1, nvim-treesitter/0ba881b7a2af0e24962e4522cf15ebd642261a90 (updated today), onedark.nvim/dac8c39812dae025255c9069a260e1f69d967927 (updated today))

> Though these don't penetrate well to lower levels, ie. `seq[Box[T]]`

That is exactly the problem. And it's not just generic types. Consider proc types with proc types in the parameters with generic types in the parameters etc. 

`type a = proc (a: proc(a: seq[seq[int]]))`

Since you can nest infinitely deep with generic types and they are just `bracket_expression`s, which contain no information about wether they are a type or not, the only clue you get is the `type_expression` at the top of the tree.

`type a = seq[seq[seq[seq[seq[seq[int]]]]]]`

```
type_section 
  type_declaration 
    type_symbol_declaration 
      name: identifier 
    type_expression 
      bracket_expression 
        left: identifier 
        right: argument_list 
          bracket_expression 
            left: identifier 
            right: argument_list 
              bracket_expression 
                left: identifier 
                right: argument_list 
                  bracket_expression 
                    left: identifier 
                    right: argument_list 
                      bracket_expression 
                        left: identifier 
                        right: argument_list 
                          bracket_expression 
                            left: identifier 
                            right: argument_list 
                              identifier 
```

AFAIK there is only 2 approaches here:
1. capture everything in `type_expression` as `@type` and overrule everything that isn't a type (what is currently happening)
2. make explicit rules for every nesting level like I did in some other cases you commented on.
Since you can't express an arbitrary amount of nesting with the query language.

Reply:
typically it's only bracket and dot expr that nests forever, but I get your point

> No. I was surprised to find out, that there seems to be no more priority based on ordering in the file. I do not know what is responsible for choosing, which rules override which, if they have no priority specified.

That's certainly interesting. Now that's gonna make a testing harness hard for me.

Right now this project uses tree-sitter-cli to test highlight queries, which have different ideas about priority (last overrides first).

Well but that's a problem for later.

Reply:
> Right now this project uses tree-sitter-cli to test highlight queries, which have different ideas about priority (last overrides first).

Yes, I'm gonna delete `test/highlight/declarations.nim` since the capture from the cli don't match with the captures from nvim and thus it breaks the CI.

Reply:
Sorry, but no. I want tests to hone in on the stuff we try to highlight.

If your queries end up requiring too many nvim quirks, then I think we should only keep a simplified query here and left the complex one for nvim-treesitter where they can test them.

Reply:
Hm. I just thought It's just 5 tests and there is no comprehensive testing going on for queries anyways at the moment, but I get your point. I guess the queries are best placed in the nvim-treesitter repo. 

I still would like to merge the change in grammar.js renaming `_type_expression` to `type_expression`

Reply:
@alaviss 

> No. I was surprised to find out, that there seems to be no more priority based on ordering in the file.

I was wrong about this.
Ordering matters for equally specific rules, also for nvim.
What I observed was a case of higher specificity. More specific rules overrule less specific ones, no matter the order.

Thus `@punctuation.bracket` will always override `@type`, when the brackets are part of the `type_expression`.
```scheme
[ "(" ")" "[" "[:" "]" "{" "}" ] @punctuation.bracket
(type_expression) @type
```